### ISSUE TITLE
Syntax highlighting of module prefix changes depending on the file content

### ISSUE THREAD

**Issue by MangelMaxime** [ionide/ionide-vscode-fsharp#1528]
## Describe the bug

See how the highlighting of the `Base.` changes when I change the file content. It should always use the same color (here the "green" color).

 
![ionide_highlight_change](https://user-images.githubusercontent.com/4760796/116432847-38906c00-a849-11eb-9166-bbb1f3bf56b3.gif)

## Code for reproduction

```fs
module Test =

    module Base =

        let fn (x : string) (y: string) =
            x + " " + y

        let fn2 (x : string) (y: string) =
            x + " " + y

    let myFund x y =
        Base.fn x y
        |> Base.fn2

    let myFund2 x y =
        Base.fn x y
        |> Base.fn2

    let myFund3 x y =
        Base.fn x

    let myFund4 x y =
        Base.fn x y

```

## Machine info
 - OS: Windows
 - .NET SDK version: 5.0.202
 - Ionide version: 5.5.3

**Comment by baronfel** [comment]
One thing that would help here is to use the `Inspect Tokens and Scopes` command from the palette and record the semantic token classes (as well as the textmate grammar classes) applied to the code in the different locations.

![image](https://user-images.githubusercontent.com/573979/116433381-0e21bd00-a80f-11eb-8648-296454c43592.png)

I have a suspicion that some of the textmate grammars are clobbering some of the semantic highlight tokens/ranges based on personal experience that I haven't turned into an issue.

**Comment by MangelMaxime** [comment]
I thought that by default VSCode used the textmate grammar, and if semantic highlight tokens is active and gives a result for the token/range it take over the textmate grammar.

I don't really like what I see from the extraction of the scope informations... When running a diff against the HTML extraction I made I get not difference...

https://www.diffchecker.com/qmdekBDU

**Information extracted when `Base` is colored in green:**

<table>
<tbody>
<tr>
    <td>language</td>	
    <td>fsharp</td>
</tr>
<tr>
    <td>standard token type</td> 
    <td>Other</td>
</tr>
<tr>
    <td>foreground</td>	
    <td>#4EC9B0</td>
</tr>
<tr>
    <td>background</td>	
    <td>#1E1E1E</td>
</tr>
<tr>
    <td>contrast ratio</td> 
    <td>8.18</td>
</tr>
<tr>
    <td>semantic token type</td> 
    <td>namespace</td>
</tr>
<tr>
    <td>foreground</td>
    <td>entity.name.namespace</td>
</tr>
<tr>
    <td></td>
    <td>
        <ul>
            <li>meta.return-type</li>
            <li>support.class</li>
            <li>support.type</li>
            <li>entity.name.type</li>
            <li>entity.name.namespace</li>
            <li>entity.other.attribute</li>
            <li>entity.name.scope-resolution</li>
            <li>entity.name.class</li>
            <li>storage.type.numeric.go</li>
            <li>storage.type.byte.go</li>
            <li>storage.type.boolean.go</li>
            <li>storage.type.string.go</li>
            <li>storage.type.uintptr.go</li>
            <li>storage.type.error.go</li>
            <li>storage.type.rune.go</li>
            <li>storage.type.cs</li>
            <li>storage.type.generic.cs</li>
            <li>storage.type.modifier.cs</li>
            <li>storage.type.variable.cs</li>
            <li>storage.type.annotation.java</li>
            <li>storage.type.generic.java</li>
            <li>storage.type.java</li>
            <li>storage.type.object.array.java</li>
            <li>storage.type.primitive.array.java</li>
            <li>storage.type.primitive.java</li>
            <li>storage.type.token.java</li>
            <li>storage.type.groovy</li>
            <li>storage.type.annotation.groovy</li>
            <li>storage.type.parameters.groovy</li>
            <li>storage.type.generic.groovy</li>
            <li>storage.type.object.array.groovy</li>
            <li>storage.type.primitive.array.groovy</li>
            <li>storage.type.primitive.groovy</li>
        </ul>
</td>
</tr>
<tr>
    <td></td>
    <td>{ "foreground": "#4EC9B0" }</td>
</tr>
<tr>
    <td>textmate token</td>	
    <td>········Base.fn·x·y (19)</td>
</tr>
<tr>    
    <td>textmate scopes</td>	
    <td>source.fsharp.fsl</td>
</tr>
</tbody>
</table>

**Information extracted when `Base` is colored in white:**

<table>
<tbody>
<tr>
    <td>language</td>	
    <td>fsharp</td>
</tr>
<tr>
    <td>standard token type</td> 
    <td>Other</td>
</tr>
<tr>
    <td>foreground</td>	
    <td>#4EC9B0</td>
</tr>
<tr>
    <td>background</td>	
    <td>#1E1E1E</td>
</tr>
<tr>
    <td>contrast ratio</td> 
    <td>8.18</td>
</tr>
<tr>
    <td>semantic token type</td> 
    <td>namespace</td>
</tr>
<tr>
    <td>foreground</td>
    <td>entity.name.namespace</td>
</tr>
<tr>
    <td></td>
    <td>
        <ul>
            <li>meta.return-type</li>
            <li>support.class</li>
            <li>support.type</li>
            <li>entity.name.type</li>
            <li>entity.name.namespace</li>
            <li>entity.other.attribute</li>
            <li>entity.name.scope-resolution</li>
            <li>entity.name.class</li>
            <li>storage.type.numeric.go</li>
            <li>storage.type.byte.go</li>
            <li>storage.type.boolean.go</li>
            <li>storage.type.string.go</li>
            <li>storage.type.uintptr.go</li>
            <li>storage.type.error.go</li>
            <li>storage.type.rune.go</li>
            <li>storage.type.cs</li>
            <li>storage.type.generic.cs</li>
            <li>storage.type.modifier.cs</li>
            <li>storage.type.variable.cs</li>
            <li>storage.type.annotation.java</li>
            <li>storage.type.generic.java</li>
            <li>storage.type.java</li>
            <li>storage.type.object.array.java</li>
            <li>storage.type.primitive.array.java</li>
            <li>storage.type.primitive.java</li>
            <li>storage.type.token.java</li>
            <li>storage.type.groovy</li>
            <li>storage.type.annotation.groovy</li>
            <li>storage.type.parameters.groovy</li>
            <li>storage.type.generic.groovy</li>
            <li>storage.type.object.array.groovy</li>
            <li>storage.type.primitive.array.groovy</li>
            <li>storage.type.primitive.groovy</li>
        </ul>
</td>
</tr>
<tr>
    <td></td>
    <td>{ "foreground": "#4EC9B0" }</td>
</tr>
<tr>
    <td>textmate token</td>	
    <td>·Base.fn2 (9)</td>
</tr>
<tr>    
    <td>textmate scopes</td>	
    <td>source.fsharp.fsl</td>
</tr>
</tbody>
</table>

**Comment by baronfel** [comment]
I've thought the same about the precedence involved, but given that the semantic tokens scopes are the same for both use cases I've not been able to come to another reason why the display would be different :-/

**Comment by MangelMaxime** [comment]
Unless the inspect token debug feature doesn't provides us will all the information for me it seems to lead to a bug in VSCode perhaps?

I mean if they are the same textmate grammar token and semantic token type it should gives the same results.

Do you know where all the `storage.type.*` stuff are coming from? I don't remember seeing them in the Ionide grammar when working on it.

**Comment by baronfel** [comment]
I don't know what that long list of `storage.type` entries are, but I have them as well. I guess they might be potentially-applicable textmate scopes?

**Comment by MangelMaxime** [comment]
I had the idea to disable the syntax semantic feature of VSCode to check what the grammar looks like by default.

![image](https://user-images.githubusercontent.com/4760796/116441380-b22c5800-a851-11eb-8c07-ff7ccfec233f.png)

We can see that the grammar never colored any text in "green" even the `module Base` and `namespace Base` are not colored. So I think we can assume that the problem is with the syntax semantic.

And the long list of items is also coming from it.

*The good news for me, is that I seem to avoid a journey into regex world 🤣*

**Comment by MangelMaxime** [comment]
The long list of `storage.type` thing seems to comes from the theme definition:

https://github.com/microsoft/vscode/blob/26ce025594734550533adf481c0509898dfd2e89/extensions/theme-defaults/themes/dark_plus.json#L21-L55

Perhaps, we see it because Ionide say I want a token of type `namespace`

I suppose the first line `foreground | entity.name.namespace` is the semantic scope chosen.

So if the scope is correct, it would leads to a rendering issue perhaps? I am not sure if we can open an issue on the VSCode repo with these informations.

Do we have an easy way to check the return of FSAC to check that FSAC gives to the correct semantic token and we understand the debug tooltip correctly?

**Comment by baronfel** [comment]
Good finds! For the semantic tokens there are a few layers involved, all of which are in FSAC, not ionide.

* Is the range reported correctly in the underlying FCS `GetSemanticRange` call? [Usage](https://github.com/fsharp/FsAutoComplete/blob/master/src/FsAutoComplete.Core/Commands.fs#L1124)
* Is the correct data still retained after we do our [custom logic](https://github.com/fsharp/FsAutoComplete/blob/master/src/FsAutoComplete.Core/Commands.fs#L98-L121) in FSAC to split out multi-line and nested ranges into a flat list of un-nested semantic tokens?
* Assuming the source data is still good, is it still correct when it comes out of the LSP-required [semantic highlight range encoding algorithm](https://github.com/fsharp/FsAutoComplete/blob/master/src/FsAutoComplete/LspHelpers.fs#L792-L836)? Or is there a bug in this algorithm?
* Can a [test case](https://github.com/fsharp/FsAutoComplete/blob/master/test/FsAutoComplete.Tests.Lsp/CoreTests.fs#L512-L519) be added to detect the error you're seeing?

This is what I'd be working through to root-cause this issue.

**Comment by baronfel** [comment]
Here's a more minimal example (at least on my system) that triggers the invalid coloration of the module name: 


```fsharp
module Base =
    let fn (x : string) (y: string) =
        x + " " + y

let myFund x y =
    Base.fn x y
```

**Comment by baronfel** [comment]
I figured out the root cause and will send an FSAC PR with tests shortly.

The bug was in step 2 listed in my comment above.  FCS understands semantic classification spans that can be nested, and so does the LSP spec, but VSCode specifically does not so instead we have logic in FSAC to split out any tokens that are nested into a set of sub-tokens that ensure a non-overlapping set of ranges for the tokens.  This could result in tokens with zero length, when the start or end of the 'parent' token was the same as the matching start or end of the 'child' token. Simply filtering these cases out ensures that coloration remains consistent as we expect.


![image](https://user-images.githubusercontent.com/573979/116629362-a8205d00-a916-11eb-8e92-3b51da5eb89e.png)

**Comment by MangelMaxime** [comment]
Awesome <3 

So much color it's looks beautiful 😍 ahah

**Comment by baronfel** [comment]
Released in 5.5.4 this afternoon.
